### DISCUSSION TITLE
'Page' object has no attribute 'insert_htmlbox'

### DISCUSSION THREAD

**Discussion by conql** [pymupdf/PyMuPDF#2951] [Q&A] [answered]
### Description of the bug

I have updated to PyMuPDF 1.23.8. But I still can't use 'insert_htmlbox'.

![image](https://github.com/pymupdf/PyMuPDF/assets/49243542/1a4e9688-a155-4141-8b22-5427da8c55a6)

### How to reproduce the bug

python -m pip install PyMuPDF --upgrade

### PyMuPDF version

1.23.8

### Operating system

Windows

### Python version

3.11

**Comment by JorjMcKie** [comment, marked as answer]
This new method has only been implemented in the new "rebased" implementation of PyMuPDF.
For a short timeframe (3 more weeks or so), please use `import fitz_new as fitz` to access it.

In first half of January we plan to switch activation defaults, so the usual import statement will point to rebased.

**Comment by JorjMcKie** [comment]
Please let me know what you think about this new feature ... also feel to ask questions or voice any suggestion.

**Comment by conql** [comment]
> Please let me know what you think about this new feature ... also feel to ask questions or voice any suggestion.

I think this is a very neat feature! Before this, insert_text was barely usable for me. Mixing Chinese and English text requires Chinese fonts, which can cause large and ugly spacing for ASCII characters. Now, with typesetting offloaded to HTML, it's much more convenient.

I was trying to use this feature to insert invisible texts into scanned PDFs to make them searchable. For my use case, however, this method is too slow, taking hundreds of times longer than the original. And most importantly, it seems to not support the alpha channel?

**Comment by JorjMcKie** [comment]
Thanks a lot for your feedback. Very reassuring.
I am gladly taking the opportunity to give some technical background on how this is implemented:

Multiple PyMuPDF components are used:

1. Story and DocumentWriter objects do the writing to an internal temporary PDF with one page that has the dimension of the target rectangle. The code in Story (which mainly is MuPDF C-code) does all the actual text output, including HTML parsing, font switching as required and calling the HarfBuzz library for text shaping.
2. If the text doesn't fit and the caller did not prohibit it, the previous step is executed multiple times. This is an iteration method that varies the scaling factor from 1 down to an optimized lower value.
3. Then, that temporary page is inserted in the original rectangle on the target page using `Page.show_pdf_page()`.

Step 1 above certainly is the most time consuming one - especially when executed numerous times because there was no initial fit. Our measurements indicate that in this case 10 to 15 executions of step 1 can occur to find an optimal scaling factor.

In comparison, `Page.insert_textbox()` and `Page.insert_text()` are basically string manipulation in Python code. No font switching, constant text properties (font size, text color, etc.). The created string is finally converted to a `bytes` object that is added (appended or prepended) to the page's `/Contents` array.

In fact, `Page.insert_htmlbox()` does not support creation of invisible text - i.e. no support of `render_mode` or `opacity` parameters.
The method however has the overlay parameter: `overlay=False` will put the text behind any other content.

As a general recommendation, you should consider using `doc.subset_fonts()` before saving. Especially if you use the typically large fonts that support Asian language, the file size savings can be very significant.
But of course you must also use garbage=3 or 4 and deflate=True. Short form `doc.ez_save(...)`.

**Reply by conql** [reply]
Thanks a ton for the detailed breakdown! I tried out overlay=False and doc.subset_fonts(), and they work like a charm. Really appreciate the effort you're putting into PyMuPDF. It is super helpful!🤗
